### DISCUSSION TITLE
Building plugins for the web development

### DISCUSSION THREAD

**Discussion by emadbaqeri** [helix-editor/helix#5328] [Ideas]
Hi folks,
As someone that has been using VIM for years I just played around with helix-editor after that moment I'm just into it and now it became my daily driver for development. While I'm doing Rust or maybe on the codebases written on go. It's awesome, but while I started a new open-source project written with typescript and React.js I have some issues with auto-completion and other helper stuff that would make my life easier. I think I can work on extensions for the web development side. One of the things that made me open this discussion thread is the CSS autocompletion which I love to have under my belt. It would save hours per month.
Any Idea how to start developing extensions or plugins for the `helix`? I should mention that it's my first to develop any tools for IDEAs and text editors. Let me know if we have such thing or can you please help me doing this?

**Comment by David-Else** [comment]
There is no plugin system yet, but it is being worked on. You can add features to Helix with pull requests written in Rust, but some things would be best as plugins, best to ask before working on something.

Regarding CSS auto-completion, TypeScript and React.js, could you be more explicit about what you are missing? Maybe it is possible now.

**Reply by emadbaqeri** [reply]
Hi @David-Else, 
Thanks for the reply, the exact thing that I'm looking for is the plugin system to build new features on top of it for the Helix. 
As I mentioned I need to work with `styled-components` inside an enormous `react` based project. I need lots of components to style and I use the `CSS-in-js` pattern with the aid of `styled-components` but unfortunately, I'm not getting any CSS completions while I'm coding the styles inside the Helix editor. Any Ideas on this?

**Comment by pascalkuthe** [comment]
Generally auto completions etc. are not implemented in helix and also not what a future plugin syatem is intended for.

Language specific functionality is usually provided by language server trough the langue server protocol (LSP). For example for rust we use rust analyzer. For CSS we use the vscode-css L's by default which you would need to.install first.

If you are unhappy with the functionality available for some language it would make more sense to contribute to the language server rather then helix

**Reply by emadbaqeri** [reply]
Hi @pascalkuthe,
Thanks for the reply. I have followed up and set up the helix as you've mentioned. Any ideas on how can I make the `styled-components` completions possible for the helix-editor?

**Reply by estin** [reply]
Hi @emadbaqeri, for completion currently you can build helix on this PRs
 -  #4816 works when lsp not available for document
 -  #3328 works with lsp (merge results) and without lsp. It is my PR to enable general words completion. Works for me around 5 month, and Helix my primary editor for stack (django  + js + rust + sql + html templates + md). Periodically synced with master

**Reply by emadbaqeri** [reply]
Hey @estin<
Thanks for the reply, I have some problems with the helix that makes me not choose it as my primary editor. I'm planning to build something on top of the helix for web-development.

**Comment by gabydd** [comment]
I wonder if this works https://github.com/styled-components/typescript-styled-plugin

**Reply by David-Else** [reply]
It looks like you can add plugins to typescript-language-server that Helix uses, you just need to work out how to add the config: https://github.com/typescript-language-server/typescript-language-server#initializationoptions

| Setting           | Type     | Description                                                                                                                                                                                                                                                          |
|:------------------|:---------|:---------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------|
| plugins           | object[] | An array of `{ name: string, location: string }` objects for registering a Typescript plugins. **Default**: []
